code-challenge-android-kotlin/app/src/main/java/com/arctouch/codechallenge/ui/HomeActivity.kt

```diff
@@ -0,0 +1,105 @@
+import com.arctouch.codechallenge.home.HomeAdapter
+import com.arctouch.codechallenge.model.Genre
+import com.arctouch.codechallenge.model.Movie
+import com.arctouch.codechallenge.webservice.MoviesWS
+import kotlinx.android.synthetic.main.home_activity.*
+
+class HomeActivity : AppCompatActivity() {
+    private var isScrolling = false
+    private var currentItems = 0
+    private var scrolledOutItems = 0
+    private var totalItems = 0
+    private var upcomingMovies = ArrayList<Movie>()
+    private var pageNumber = 1L
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.home_activity)
+
+        val viewModel = ViewModelProviders.of(this).get(HomeViewModel::class.java)
+
+        viewModel.getGenres().observe(this, Observer<List<Genre>>{genres ->
+            viewModel.getUpcomingMovies(pageNumber).observe(this, Observer<List<Movie>>{ movies ->
+                movies?.let{
+                    upcomingMovies = ArrayList(movies)
+                    val layoutManager = LinearLayoutManager(this)
+                    val adapter = HomeAdapter(this, upcomingMovies, showMovieFragmentLayout,
+                    hideMovieFragmentLayout, showProgressBar, hideProgressBar)
+                    recyclerView.layoutManager = layoutManager
+                    recyclerView.adapter = adapter
+                    val scrollListener = getScrollListener(layoutManager, adapter)
+                    recyclerView.addOnScrollListener(scrollListener)
+                    progressBar.visibility = View.GONE
+            }
+
+                //TODO notify user if movies = null
+            })
+        })
+    }
+
+    var showMovieFragmentLayout: () -> Unit = {
+        frameLayout.visibility = View.VISIBLE
+    }
+
+    var hideMovieFragmentLayout: () -> Unit = {
+        frameLayout.visibility = View.GONE
+    }
+
+    var showProgressBar: () -> Unit = {
+        progressBar.visibility = View.VISIBLE
+    }
+
+    var hideProgressBar: () -> Unit = {
+        progressBar.visibility = View.GONE
+    }
+
+    fun getScrollListener(layoutManager: LinearLayoutManager, adapter: HomeAdapter) : RecyclerView.OnScrollListener {
+        return object: RecyclerView.OnScrollListener() {
+            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
+                super.onScrolled(recyclerView, dx, dy)
+                currentItems = layoutManager.childCount
+                scrolledOutItems = layoutManager.findFirstVisibleItemPosition()
+                totalItems = layoutManager.itemCount
+
+                if(isScrolling && (currentItems + scrolledOutItems == totalItems)){
+                    isScrolling = false
+                    pageNumber++
+                    progressBar.visibility = View.VISIBLE
+                    //TODO move api call to viewmodel
+                    MoviesWS().getUpcomingMovies(pageNumber, {
```

Comment:
Here you're calling `MoviesWS` inside the View (Activity). According to the MVVM, this should be concern of the ViewModel or even a Repository to load the data and then pass to the ViewModel to update the UI that is observing model changes. This links has a lot of blueprint on how we can thrive using Architecture Components: https://github.com/googlesamples/android-architecture-components

Reply:
Also read more about `God Object` or `God Activity` anti-pattern and how we can avoid them: https://medium.com/spring-media-techblog/keep-your-droid-clean-activities-and-fragments-da1799fe1e7a

Reply:
My plan was to move MovieWS to the ViewModel, as there was a method responsible to call it already implemented there.  I started the task coding the scrollListener first and I thought I was going to need a more elaborated approach to call the ViewModel from the ScrollListener and update the RecyclerView so I decided to do it like that to test it first. It turned out I didn't have time to move it. Looking at the code now, I noticed I needed little effort to do it!

Reply:
Oh, and thank you for the links! :D I'll sure be reading them as soon as possible.